### PR TITLE
Fix #17052 - Add support for Keycloak Authentication

### LINKED ISSUE #17052: Add support to login with Keycloak
### Is your feature request related to a problem? Please describe.

In our cluster with Kubernetes, we are using Keycloak (https://www.keycloak.org/) to log in to our applications using our Google accounts. It will be interesting if there is support for this Authentication. Keycloak authentication is using JWT (https://jwt.io/) token.

### Describe the solution you'd like

It will be interesting if we could use our Keycloak authentication to log in to the PHP my admin page. A solution for this could be having a PHP class extending from the abstract class AuthenticationPlugin.php.

### Additional context

We are using the latest version of phpMyAdmin.

### PR DESCRIPTION
**Pull request by davimenezesoli**
This PR will fix the issue: https://github.com/phpmyadmin/phpmyadmin/issues/17052

Signed-off-by: Davi Menezes <davi.menezes@visma.com>

### Description

Please describe your pull request.

Fixes #17052

Before submitting pull request, please review the following checklist:

- [x] Make sure you have read our [CONTRIBUTING.md](https://github.com/phpmyadmin/phpmyadmin/blob/master/CONTRIBUTING.md) document.
- [x] Make sure you are making a pull request against the correct branch. For example, for bug fixes in a released version use the corresponding QA branch and for new features use the _master_ branch. If you have a doubt, you can ask as a comment in the bug report or on the mailing list.
- [x] Every commit has proper `Signed-off-by` line as described in our [DCO](https://github.com/phpmyadmin/phpmyadmin/blob/master/DCO). This ensures that the work you're submitting is your own creation.
- [x] Every commit has a descriptive commit message.
- [x] Every commit is needed on its own, if you have just minor fixes to previous commits, you can squash them.
- [x] Any new functionality is covered by tests.

### CODE REVIEW

File: libraries/classes/Plugins/Auth/AuthenticationKeycloak.php [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+class AuthenticationKeycloak extends AuthenticationPlugin
+{
+    public function showLoginForm(): bool
+    {
+        return false;
+    }
+
+    public function readCredentials(): bool
+    {
+        if (!isset($_COOKIE["kc-access"])) {
```
Comment (williamdes) [inline libraries/classes/Plugins/Auth/AuthenticationKeycloak.php:24]:
```suggestion
        if (!isset($_COOKIE['kc-access'])) {
```
`composer run phpcbf` will help you

File: libraries/classes/Plugins/Auth/AuthenticationKeycloak.php [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        $this->user = $token->getClaim('preferred_username');
+        $this->password = $token->getClaim('sub');
+
+        return true;
+    }
+
+    /**
+     * @param string $keycloakToken
+     * @return Token
+     */
```
Comment (williamdes) [inline libraries/classes/Plugins/Auth/AuthenticationKeycloak.php:47]: This one can be removed as it adds nothing to existing php typehints

File: libraries/classes/Plugins/Auth/AuthenticationKeycloak.php [hunk 3]
```diff
@@ -0,0 +1,53 @@
+<?php
+
+declare(strict_types=1);
+
+namespace PhpMyAdmin\Plugins\Auth;
+
+use Lcobucci\JWT\Parser;
```
Comment (williamdes) [inline libraries/classes/Plugins/Auth/AuthenticationKeycloak.php:7]: This would not work without adding the lib to composer.json

**Comment by williamdes** [comment]
Hi @mauriciofauth 
Is it time to think about the plugin system for this ?

**Comment by mauriciofauth** [comment]
> Hi @mauriciofauth 
> Is it time to think about the plugin system for this ?

Definitely. I already have an idea. I will work on that.

File: test/classes/Plugins/Auth/AuthenticationKeycloakTest.php [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+            $this->object->readCredentials()
+        );
+
+        unset($_COOKIE['kc-access']);
+    }
+
+    public function testReadCredentialsWithoutAllFields()
+    {
+        $_COOKIE['kc-access'] = 'unit-test';
+        $token = Mockery::mock(Token::class);
```
Comment (mauriciofauth) [inline test/classes/Plugins/Auth/AuthenticationKeycloakTest.php:77]: Could you replace Mockery with [PHPUnit's mock](https://phpunit.readthedocs.io/en/8.5/test-doubles.html#mock-objects)?

**Review by MauricioFauth — CHANGES_REQUESTED** [review]

File: test/classes/Plugins/Auth/AuthenticationKeycloakTest.php (hunk 4 above)
Comment (davimenezesoli) [reply test/classes/Plugins/Auth/AuthenticationKeycloakTest.php]: I updated the tests to not mock it.

File: test/classes/Plugins/Auth/AuthenticationKeycloakTest.php [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+            $this->object->readCredentials()
+        );
+
+        unset($_COOKIE['kc-access']);
+    }
+
+    public function testReadCredentialsWithoutAllFields()
+    {
+        // This is a public token for test purposes
+        $_COOKIE['kc-access'] = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_adQssw5c';
```
Comment (williamdes) [inline test/classes/Plugins/Auth/AuthenticationKeycloakTest.php:75]: it has no secret configured somewhere ?

Comment (davimenezesoli) [reply test/classes/Plugins/Auth/AuthenticationKeycloakTest.php:75]: No, this JWT token is set in the cookie and we read from it to authenticate the user or fail.

Comment (williamdes) [reply test/classes/Plugins/Auth/AuthenticationKeycloakTest.php:75]: A jwt without a secret to verify it makes not sense to me
But it is maybe okay here as we decode the payload to get the data, but in one sense we can not trust the cookie value

**Pull request closed without merging**